examples/lessons/13/app.js

```diff
@@ -0,0 +1,294 @@
+
+varying vec2 vTextureCoord;
+varying vec4 vTransformedNormal;
+varying vec4 vPosition;
+
+void main(void) {
+  vPosition = uMMatrix * vec4(positions, 1.0);
+  gl_Position = uPMatrix * uVMatrix * vPosition;
+  vTextureCoord = texCoords;
+  vTransformedNormal = uMMatrix * vec4(normals, 0.0);
+}
+`;
+
+const FRAGMENT_LIGHTING_FRAGMENT_SHADER = `\
+#ifdef GL_ES
+precision highp float;
+#endif
+
+uniform vec3 uAmbientColor;
+
+uniform vec3 uPointLightingLocation;
+uniform vec3 uPointLightingColor;
+
+uniform bool uUseLighting;
+uniform bool uUseTextures;
+
+varying vec2 vTextureCoord;
+varying vec4 vTransformedNormal;
+varying vec4 vPosition;
+
+uniform sampler2D uSampler;
+
+void main(void) {
+  vec3 lightWeighting;
+
+  if (!uUseLighting) {
+    lightWeighting = vec3(1.0, 1.0, 1.0);
+  } else {
+    vec3 lightDirection = normalize(uPointLightingLocation - vPosition.xyz);
+    float pointLightWeighting = max(dot(vTransformedNormal.xyz, lightDirection), 0.0);
+    lightWeighting = uAmbientColor + uPointLightingColor * pointLightWeighting;
+  }
+
+  vec4 fragmentColor;
+  if (uUseTextures) {
+    fragmentColor = texture2D(uSampler, vec2(1.0 - vTextureCoord.s, 1.0 - vTextureCoord.t));
+  } else {
+    fragmentColor = vec4(1.0, 1.0, 1.0, 1.0);
+  }
+  gl_FragColor = vec4(fragmentColor.rgb * lightWeighting, fragmentColor.a);
+}
+`;
+
+const moonRotation = {
+  matrix: new Matrix4().rotateY(radians(180)).translate([2, 0, 0])
+};
+
+const cubeRotation = {
+  matrix: new Matrix4().translate([1.25, 0, 0])
+};
+
+const animationLoop = new AnimationLoop({
+  onInitialize: ({canvas, gl}) => {
+
+    setParameters(gl, {
+      clearColor: [0, 0, 0, 1],
+      clearDepth: 1,
+      depthTest: true
+    });
+
+    let vertexLightingProgram = new Program(gl, {
+      fs: VERTEX_LIGHTING_FRAGMENT_SHADER,
+      vs: VERTEX_LIGHTING_VERTEX_SHADER,
+    });
+
+    let fragmentLightingProgram = new Program(gl, {
+      fs: FRAGMENT_LIGHTING_FRAGMENT_SHADER,
+      vs: FRAGMENT_LIGHTING_VERTEX_SHADER,
+    });
+
+    return loadTextures(gl, {
+      urls: ['moon.gif']
+    })
+    .then(textures => {
+      let moon = new Sphere(gl, {
+        program: fragmentLightingProgram,
+        uniforms: {
+          uSampler: textures[0]
+        },
+        nlat: 30,
+        nlong: 30,
+        radius: 1,
+      });
+      return loadTextures(gl, {
+        urls: ['crate.gif']
+      })
+      .then(textures => {
+        let cube = new Cube(gl, {
+          program: fragmentLightingProgram,
+          uniforms: {
+            uSampler: textures[0]
+          }
+        });
+        return {moon, cube, vertexLightingProgram, fragmentLightingProgram};
+      });
+    });
+  },
+  onRender: ({
+    gl, tick, aspect, moon, cube, vertexLightingProgram, fragmentLightingProgram
+  }) => {
+    // set camera position
+    const eyePos = new Matrix4()
+    .rotateX(radians(-30))
+    .transformVector3([0, 0, 5]);
+
+    gl.clear(GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT);
```

Comment:
Nit; I'd move this line `gl.clear` up or down to keep eyePos and uVMatrix calculations together.

Reply:
Done.